File: services/cloud_api/consult_live.py
"""
VitaScribe Cloud API - Live consult

Volgt het gesprek tussen arts en patiënt terwijl het plaatsvindt, en maakt
na "stop" het verslag. Het geluid gaat in kleine stukjes via deze server
naar Deepgram (EU), dat de stemmen scheidt (diarize). De server bewaart
alleen de tekst per spreker, in het geheugen en alleen voor dit consult.
Na "stop" gaat dat transcript door dezelfde verwerking als een opgenomen
consult (pipeline.verwerk_transcript).

Waarom live en niet achteraf uploaden: het verslag staat er een paar
seconden na "stop" in plaats van na het uploaden en uitschrijven van het
hele consult, er is geen grens aan de bestandsgrootte, en er staat nooit
een audiobestand op de server. De extensie houdt zelf een kopie van de
opname tot het verslag binnen is; valt de verbinding weg, dan stuurt zij
die kopie alsnog op de gewone manier (/api/v1/consult/process).

Tijdens het consult krijgt de arts geen meelopende tekst te zien, alleen
dat er geluisterd wordt en hoeveel stemmen er zijn gehoord. Tussentijdse
tekst is vaak nog fout en leidt af van de patiënt.

Protocol (WebSocket /api/v1/consult/stream):
  client -> server  {"type": "auth", "api_key": "...", "praktijk": "...",
                     "consent": true, "keyterms": [...], "llm_provider": ...}
  client -> server  <binaire audio, webm/opus>
  client -> server  {"type": "stop"}
  server -> client  {"type": "ready"}
  server -> client  {"type": "voortgang", "seconden": 12.3, "sprekers": 2}
  server -> client  {"type": "verwerken"}
  server -> client  {"type": "result", "data": {...}, "leeg": bool}
  server -> client  {"type": "error", "message": "...", "terugval": bool}
  server -> client  {"type": "closed"}

"terugval": true betekent: stuur de eigen opname alsnog op.
Compatible with Python 3.9+.
"""

from __future__ import annotations

import asyncio
import json
import os
import time
from typing import Any, Callable, Dict, List, Optional
from urllib.parse import urlencode

import structlog
from fastapi import WebSocket, WebSocketDisconnect

from . import audit, pipeline
from .config import AppConfig, get_config
from .dictation import (
    KEYTERM_RETRY_BUDGETS,
    MAX_KEYTERM_TOKENS,
    _authenticate,
    _default_connect,
    _send_json,
    build_keyterms,
    sanitize_user_keyterms,
)
from .stt_service import TranscriptResult, TranscriptSegment

logger = structlog.get_logger()

UPSTREAM_CLOSE_TIMEOUT_SECS = 10.0


def max_seconden() -> int:
    """Kostenrem: een vergeten microfoon mag niet uren doorlopen."""
    try:
        return int(os.getenv("CONSULT_LIVE_MAX_SECONDS", "2700"))
    except ValueError:
        return 2700


def build_consult_url(cfg: AppConfig, user_terms: Optional[List[str]] = None,
                      token_budget: int = MAX_KEYTERM_TOKENS) -> str:
    """Streaming-adres voor een gesprek: stemmen scheiden, geen tussentekst."""
    model = cfg.stt.deepgram_model
    params: List[tuple] = [
        ("model", model),
        ("language", cfg.stt.deepgram_language),
        ("punctuate", "true"),
        ("smart_format", "true"),
        ("diarize", "true"),
        ("interim_results", "false"),
        # AVG: niet bewaard en niet gebruikt voor training.
        ("mip_opt_out", "true"),
    ]
    if token_budget > 0 and cfg.dictation.keyterms_enabled and model.startswith("nova-3"):
        params.extend(("keyterm", term) for term in build_keyterms(user_terms, token_budget))
    return f"{cfg.dictation.deepgram_url}?{urlencode(params)}"


class Gesprek:
    """De tekst van het consult per spreker, opgebouwd uit Deepgram-finals."""

    def __init__(self) -> None:
        self.segmenten: List[TranscriptSegment] = []
        self.sprekers: set = set()
        self.seconden = 0.0

    def verwerk(self, raw: Any) -> bool:
        """Neemt een Deepgram-bericht op. True als er tekst bij kwam."""
        if isinstance(raw, bytes):
            return False
        try:
            data = json.loads(raw)
        except (TypeError, ValueError):
            return False
        if data.get("type") == "Metadata":
            self.seconden = max(self.seconden, float(data.get("duration") or 0.0))
            return False
        if data.get("type") != "Results" or not data.get("is_final"):
            return False
        alternatieven = (data.get("channel") or {}).get("alternatives") or [{}]
        woorden = alternatieven[0].get("words") or []
        erbij = False
        for w in woorden:
            tekst = (w.get("punctuated_word") or w.get("word") or "").strip()
            if not tekst:
                continue
            spreker = f"spreker_{w.get('speaker', 0)}"
            begin, eind = float(w.get("start") or 0.0), float(w.get("end") or 0.0)
            laatste = self.segmenten[-1] if self.segmenten else None
            if laatste is not None and laatste.speaker == spreker:
                laatste.text += " " + tekst
                laatste.end = eind
            else:
                self.segmenten.append(TranscriptSegment(text=tekst, start=begin, end=eind, speaker=spreker))
            self.sprekers.add(spreker)
            self.seconden = max(self.seconden, eind)
            erbij = True
        return erbij

    def transcript(self) -> TranscriptResult:
        return TranscriptResult(
            raw_text=" ".join(s.text for s in self.segmenten),
            segments=list(self.segmenten),
            language="nl",
            duration_secs=round(self.seconden, 1),
            provider="deepgram-live",
        )


async def volg_consult(
    ws: WebSocket,
    connect: Callable[[str, str], Any] = _default_connect,
    verwerk: Callable[..., Any] = pipeline.verwerk_transcript,
) -> None:
    """Een live consult: audio in, na stop het verslag uit."""
    await ws.accept()
    cfg = get_config()

    auth, ident = await _authenticate(ws)
    if auth is None:
        await _send_json(ws, {"type": "error", "message": ident, "terugval": False})
        await ws.close(code=4401)
        return

    # KNMG (2026): een consult opnemen mag alleen met toestemming van de patiënt.
    if os.getenv("REQUIRE_RECORDING_CONSENT", "true").lower() == "true" and auth.get("consent") is not True:
        audit.log_event(ident.label, "consult.refused", status="geen_toestemming")
        await _send_json(ws, {"type": "error", "terugval": False,
                              "message": "Toestemming van de patiënt voor de opname is niet bevestigd."})
        await ws.close(code=4400)
        return

    from fastapi import HTTPException
    from .praktijk_sleutels import kies_spraak
    try:
        deepgram_key = await kies_spraak(ident)
    except HTTPException as exc:
        await _send_json(ws, {"type": "error", "message": exc.detail, "terugval": False})
        await ws.close(code=4403)
        return
    if not deepgram_key:
        await _send_json(ws, {"type": "error", "terugval": True,
                              "message": "Spraakherkenning is niet ingesteld op de server."})
        await ws.close(code=4500)
        return

    user_terms = sanitize_user_keyterms(auth.get("keyterms"))
    upstream = None
    last_exc: Optional[Exception] = None
    for budget in KEYTERM_RETRY_BUDGETS:
        try:
            upstream = await connect(build_consult_url(cfg, user_terms, budget), deepgram_key)
            break
        except Exception as exc:
            last_exc = exc
            logger.warning("consult_live.upstream_connect_failed", error=str(exc), keyterm_budget=budget)
            if "400" not in str(exc):
                break
    if upstream is None:
        await _send_json(ws, {"type": "error", "terugval": True,
                              "message": f"Kan spraakherkenning niet bereiken: {last_exc}"})
        await ws.close(code=4502)
        return

    audit.log_event(ident.label, "consult.stream", consent=True)
    logger.info("consult_live.start", model=cfg.stt.deepgram_model)
    await _send_json(ws, {"type": "ready"})

    gesprek = Gesprek()
    gestopt = asyncio.Event()   # de arts klikte op stop (of de tijd is om)

    async def client_naar_deepgram() -> None:
        try:
            while True:
                message = await ws.receive()
                if message["type"] == "websocket.disconnect":
                    return
                if message.get("bytes"):
                    await upstream.send(message["bytes"])
                elif message.get("text"):
                    try:
                        control = json.loads(message["text"])
                    except ValueError:
                        continue
                    if control.get("type") == "stop":
                        gestopt.set()
                        break
        except WebSocketDisconnect:
            return
        try:
            await upstream.send(json.dumps({"type": "CloseStream"}))
        except Exception:
            pass

    async def deepgram_naar_gesprek() -> None:
        try:
            async for raw in upstream:
                if gesprek.verwerk(raw):
                    await _send_json(ws, {"type": "voortgang", "seconden": round(gesprek.seconden, 1),
                                          "sprekers": len(gesprek.sprekers)})
        except Exception as exc:
            logger.warning("consult_live.upstream_closed", error=str(exc))

    zender = asyncio.create_task(client_naar_deepgram())
    ontvanger = asyncio.create_task(deepgram_naar_gesprek())
    start = time.time()
    try:
        klaar, _ = await asyncio.wait({zender, ontvanger}, timeout=max_seconden(),
                                      return_when=asyncio.FIRST_COMPLETED)
        if not klaar:
            # Tijd om: stoppen en verwerken wat er is.
            zender.cancel()
            gestopt.set()
            try:
                await upstream.send(json.dumps({"type": "CloseStream"}))
            except Exception:
                pass
        elif ontvanger in klaar and not gestopt.is_set():
            # Deepgram viel weg terwijl het consult nog liep.
            zender.cancel()
            await _send_json(ws, {"type": "error", "terugval": True,
                                  "message": "De live verbinding met de spraakherkenning viel weg."})
            return

        if not gestopt.is_set():
            return   # de extensie verbrak de verbinding zonder stop: niets te doen

        try:
            await asyncio.wait_for(ontvanger, timeout=UPSTREAM_CLOSE_TIMEOUT_SECS)
        except asyncio.TimeoutError:
            ontvanger.cancel()

        await _send_json(ws, {"type": "verwerken"})
        transcript = gesprek.transcript()
        result = await verwerk(transcript, llm_provider=auth.get("llm_provider"))
        data = result.to_dict()
        data["processing_time_secs"] = round(time.time() - start, 2)
        await _send_json(ws, {"type": "result", "data": data, "leeg": not transcript.raw_text.strip()})
        logger.info("consult_live.complete", seconden=transcript.duration_secs,
                    sprekers=len(gesprek.sprekers))
    except Exception as exc:
        logger.error("consult_live.error", error=str(exc))
        await _send_json(ws, {"type": "error", "terugval": True,
                              "message": "Het verslag kon niet worden gemaakt."})
    finally:
        for taak in (zender, ontvanger):
            if not taak.done():
                taak.cancel()
        try:
            await upstream.close()
        except Exception:
            pass
        await _send_json(ws, {"type": "closed"})
        try:
            await ws.close()
        except RuntimeError:
            pass

File: chrome-extension/lib/consult-live.js
/**
 * VitaScribe - Live consult (verbinding met /api/v1/consult/stream)
 *
 * Stuurt het geluid van het consult in kleine stukjes naar de server, die
 * het gesprek per spreker volgt en na "stop" het verslag maakt. De opname
 * zelf houdt bricks.js bij als reservekopie: gaat hier iets mis, dan geeft
 * stop() { ok: false } terug en stuurt bricks.js de kopie alsnog op de
 * gewone manier op. Een consult gaat dus nooit verloren door een haperende
 * verbinding.
 *
 * Geen DOM en geen chrome.*: de WebSocket wordt meegegeven, zodat dit met
 * node --test na te lopen is.
 */
var SVConsultLive = (function () {
  'use strict';

  var OPEN = 1;

  /**
   * opties: { apiUrl, apiKey, praktijk, llmProvider, WebSocket,
   *           onVoortgang(seconden, sprekers), onFout(melding, terugval) }
   */
  function start(opties) {
    var WS = opties.WebSocket || (typeof WebSocket !== 'undefined' ? WebSocket : null);
    var adres = String(opties.apiUrl || '').replace(/\/$/, '').replace(/^http/, 'ws') + '/api/v1/consult/stream';
    var ws = new WS(adres);
    var klaar = false;       // server gaf "ready"
    var mislukt = false;     // live werkt niet meer; bricks.js valt terug
    var stoppen = false;
    var wachtrij = [];
    var afronden = null;     // resolve van stop()
    var timer = null;
    var uitkomst = null;

    function einde(resultaat) {
      if (uitkomst) return;
      uitkomst = resultaat;
      clearTimeout(timer);
      if (afronden) afronden(resultaat);
      try { if (ws.readyState <= OPEN) ws.close(); } catch (e) { /* al dicht */ }
    }

    function faal(melding, terugval) {
      if (mislukt) return;
      mislukt = true;
      wachtrij = [];
      if (stoppen) einde({ ok: false, terugval: terugval !== false, melding: melding });
      else if (opties.onFout) opties.onFout(melding, terugval !== false);
    }

    function verstuur(data) {
      if (mislukt) return;
      if (klaar && ws.readyState === OPEN) ws.send(data);
      else wachtrij.push(data);
    }

    ws.onopen = function () {
      ws.send(JSON.stringify({ type: 'auth', api_key: opties.apiKey || '', praktijk: opties.praktijk || '',
                               consent: true, llm_provider: opties.llmProvider || null }));
    };
    ws.onmessage = function (bericht) {
      var e;
      try { e = JSON.parse(bericht.data); } catch (err) { return; }
      if (e.type === 'ready') {
        klaar = true;
        var rij = wachtrij;
        wachtrij = [];
        rij.forEach(function (d) { ws.send(d); });
      } else if (e.type === 'voortgang') {
        if (opties.onVoortgang) opties.onVoortgang(e.seconden || 0, e.sprekers || 0);
      } else if (e.type === 'result') {
        if (e.leeg) einde({ ok: false, terugval: true, melding: 'Live werd geen spraak gehoord.' });
        else einde({ ok: true, data: e.data });
      } else if (e.type === 'error') {
        faal(e.message || 'Fout in het live consult.', e.terugval);
      }
    };
    ws.onerror = function () { faal('De live verbinding met de server viel weg.', true); };
    ws.onclose = function () {
      if (!uitkomst) {
        if (stoppen && mislukt) einde({ ok: false, terugval: true, melding: 'De live verbinding werd gesloten.' });
        else faal('De live verbinding werd gesloten.', true);
      }
    };

    return {
      /** Een stuk geluid van de recorder. */
      stuur: verstuur,
      /** Werkt live nog? Zo niet, dan loopt de opname alleen lokaal door. */
      gezond: function () { return !mislukt; },
      /**
       * Na de laatste audio: vraag het verslag. Levert { ok, data } of
       * { ok: false, terugval, melding }.
       */
      stop: function (wachtMs) {
        stoppen = true;
        return new Promise(function (resolve) {
          afronden = resolve;
          if (uitkomst) { resolve(uitkomst); return; }
          if (mislukt) { einde({ ok: false, terugval: true, melding: 'Live verbinding was al weggevallen.' }); return; }
          verstuur(JSON.stringify({ type: 'stop' }));
          timer = setTimeout(function () {
            einde({ ok: false, terugval: true, melding: 'Het live verslag bleef te lang weg.' });
          }, wachtMs || 90000);
        });
      },
      /** Afbreken zonder verslag. */
      sluit: function () { einde({ ok: false, terugval: false, melding: 'Afgebroken.' }); },
    };
  }

  return { start: start };
})();

if (typeof module !== 'undefined' && module.exports) module.exports = SVConsultLive;
